packages/app-store/nextcloudtalk/lib/VideoApiAdapter.ts

```diff
@@ -0,0 +1,209 @@
+import { stringify } from "querystring";
+import { v4 as uuidv4 } from "uuid";
+import { z } from "zod";
+
+import logger from "@calcom/lib/logger";
+import { safeStringify } from "@calcom/lib/safeStringify";
+import prisma from "@calcom/prisma";
+import type { CalendarEvent } from "@calcom/types/Calendar";
+import type { CredentialPayload } from "@calcom/types/Credential";
+import type { PartialReference } from "@calcom/types/EventManager";
+import type { VideoApiAdapter, VideoCallData } from "@calcom/types/VideoApiAdapter";
+
+import getAppKeysFromSlug from "../../_utils/getAppKeysFromSlug";
+import getParsedAppKeysFromSlug from "../../_utils/getParsedAppKeysFromSlug";
+import { invalidateCredential } from "../../_utils/invalidateCredential";
+import { OAuthManager } from "../../_utils/oauth/OAuthManager";
+import { markTokenAsExpired } from "../../_utils/oauth/markTokenAsExpired";
+import { oAuthManagerHelper } from "../../_utils/oauth/oAuthManagerHelper";
+import config from "../config.json";
+import { appKeysSchema } from "../zod";
+
+const nextcloudEventResultSchema = z.object({
+  ocs: z.object({
+    meta: z.object({
+      status: z.string(),
+      statuscode: z.number(),
+      message: z.string(),
+    }),
+    data: z.object({
+      token: z.string(),
+    }),
+  }),
+});
+
+const NextcloudTalkVideoApiAdapter = (credential: CredentialPayload): VideoApiAdapter => {
+  const tokenResponse = oAuthManagerHelper.getTokenObjectFromCredential(credential);
+
+  const clientCredentials = getParsedAppKeysFromSlug(config.slug, appKeysSchema);
+
+  const fetchNextcloudApi = async (endpoint: string, options?: RequestInit) => {
+    const auth = new OAuthManager({
+      credentialSyncVariables: oAuthManagerHelper.credentialSyncVariables,
+      resourceOwner: {
+        type: "user",
+        id: credential.userId,
+      },
+      appSlug: config.slug,
+      currentTokenObject: tokenResponse,
+      fetchNewTokenObject: async ({ refreshToken }: { refreshToken: string | null }) => {
+        const { nextcloudTalkClientId, nextcloudTalkClientSecret, nextcloudTalkHost } =
+          await clientCredentials;
+        if (!refreshToken) {
+          return null;
+        }
+        const params = {
+          client_id: nextcloudTalkClientId,
+          client_secret: nextcloudTalkClientSecret,
+          refresh_token: refreshToken,
+          grant_type: "refresh_token",
+        };
+        const query = stringify(params);
+        return fetch(`${nextcloudTalkHost}/index.php/apps/oauth2/api/v1/token?${query}`, {
```

Comment:
Occasionally, while testing I had HTTP 500 responses from my Nextcloud API on the first time of calling this endpoint. When I then called it for the second time (exact same call and params) it just worked. This also happened when I used my local API client for testing which makes me assume that my Nextcloud instance is the issue and not this integration. Also, it only happened sometimes. So, I could not consistently reproduce that behavior.

However, what I did locally was to add a `fetchRetry.js` file in `packages/app-store/_utils` with this code:

```js
async function fetchRetry(input: RequestInfo | URL, init?: RequestInit) {
  let count = 3;
  while (count > 0) {
    try {
      return await fetch(input, init);
    } catch (error) {}

    count -= 1;
  }

  throw new Error(`Too many retries`);
}

export default fetchRetry;
```

So when importing `fetchRetry` this line would become
```suggestion
        return fetchRetry(`${nextcloudTalkHost}/index.php/apps/oauth2/api/v1/token?${query}`, {
```

Like this I never had failed API requests. However, as it's probably just my Nextcloud instance I'm unsure if I should commit this. It's would be just a precaution. What do you think about that @Amit91848 ?